applications/luci-app-openvpn/luasrc/controller/openvpn.lua

```diff
@@ -8,4 +8,47 @@ function index()
 	entry( {"admin", "services", "openvpn"}, cbi("openvpn"), _("OpenVPN") )
 	entry( {"admin", "services", "openvpn", "basic"},    cbi("openvpn-basic"),    nil ).leaf = true
 	entry( {"admin", "services", "openvpn", "advanced"}, cbi("openvpn-advanced"), nil ).leaf = true
+	entry( {"admin", "services", "openvpn", "file"},     form("openvpn-file"),    nil ).leaf = true
+	entry( {"admin", "services", "openvpn", "upload"},   call("ovpn_upload"))
+end
+
+function ovpn_upload()
+	local fs     = require("nixio.fs")
+	local http   = require("luci.http")
+	local util   = require("luci.util")
+	local uci    = require("luci.model.uci").cursor()
+	local upload = http.formvalue("ovpn_file")
+	local name   = http.formvalue("instance_name2")
+	local file   = "/etc/openvpn/" ..name.. ".ovpn"
```

Comment:
The `name` value should be thoroughly validated here. Imagine it containing something like `../passwd%00`

Reply:
The instance_name2 will be already filtered by client side javascript in "cbi-select-input-add.htm":
```
var vpn_name = div_upload.querySelector("#instance_name2").value.replace(/[^\x00-\x7F]|[\s!@#$%^&*()+=\[\]{};':"\\|,<>\/?]/g,'');
```
Nevertheless, I've added the shellquote function on lua side, too.